src/runtime/GlobalObjectBuiltinString.cpp

```diff
@@ -886,7 +878,9 @@ static Value builtinStringStartsWith(ExecutionState& state, Value thisValue, siz
     // Let isRegExp be ? IsRegExp(searchString).
     // If isRegExp is true, throw a TypeError exception.
     if (searchString.isObject() && searchString.asObject()->isRegExpObject()) {
-        ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, "can't use RegExp with startsWith");
+        if (searchString.asObject()->asRegExpObject()->checkMatchSymbol(state)) {
+            ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, "can't use RegExp with startsWith");
```

Comment:
Is it possible that `searchString` is an Object other than RegExpObject and also has @@match property? In that case, [`isRegExp`](https://www.ecma-international.org/ecma-262/6.0/#sec-isregexp) might be true and we should throw a TypeError exception here too.

Reply:
The standard only specifies at startsWith:
```
Let isRegExp be IsRegExp(searchString).
ReturnIfAbrupt(isRegExp).
If isRegExp is true, throw a TypeError exception.
```

while you copied the isRegex above, which does not care about the exact type, only if the parameter is an object. Also It seems like the @@match symbol is not present, if the object is not a RegExp object itself. Tested with node and mdn with this little example:

```js
var otherObject = 4;
otherObject[Symbol.match] = false;
console.log(Object.getOwnPropertySymbols(otherObject))
console.log(otherObject instanceof RegExp)

var rx = /foo/
rx[Symbol.match] = false;
console.log(Object.getOwnPropertySymbols(rx))
console.log(rx instanceof RegExp)
```  
returns:  
```
> Array []
> false
> Array [Symbol(Symbol.match)]
> true
```
it is because at `otherObject[Symbol.match] = false;` then the otherObject is coerced to a "fake object" which will have a @@match symbol, but that is not set back. As far as i know, you cannot set @@match to a non-regex object, as it will not have the symbol by default.

```js
var rx = /foo/
console.log(rx[Symbol.match])

var dummy = 4
console.log(dummy[Symbol.match])
```
results:
```
> function [Symbol.match]() { [native code] }
> undefined
```

---
Just to test with true @@match

```js
var otherObject = 4;
otherObject[Symbol.match] = true;
console.log(Object.getOwnPropertySymbols(otherObject))
console.log("4asd".startsWith(otherObject))
console.log(otherObject instanceof RegExp)
```
returns:
```
> Array []
> true
> false
```

which correlates to the current working of the patch.

Reply:
@Achie72 please consider another example as follow.
```js
var search = new String("to");
search[Symbol.match] = true;
console.log(search[Symbol.match]); // true

var str = "tool";
str.startsWith(search); // TypeError: First argument to String.prototype.startsWith must not be a regular expression
```
`search` is a String Object which also has `Symbol.match` property.
If `search` does not have the Symbol property, then the result of `str.startsWith` should be true.
However TypeError exception is triggered because `IsRegExp(search)` returns true value from the standard.

> IsRegExp (search)
    1. If Type(argument) is not Object, return false.
    2. Let isRegExp be Get(argument, @@match).
    // isRegExp is true by `search[Symbol.match] = true`
    3. ReturnIfAbrupt(isRegExp).
    4. **If isRegExp is not undefined, return ToBoolean(isRegExp).** 
    // isRegExp is a true value, so true is returned at this point
    5. If argument has a [[RegExpMatcher]] internal slot, return true.
    6. Return false.

Reply:
@clover2123 i updated the PR, could you take a look at it?

As we don't have a seperate isRegExp i merged the `If Type(argument) is not Object, return false.` and the `isRegExp is true by search[Symbol.match] = true` into the if statement, which i think should handle these cases correctly. As this does not defines an `the object need to be a RegExp` i removed the `isRegexObject` check, and solely rely on the upper two conditions.